operating_unit/models/operating_unit_mixins.py

```diff
@@ -0,0 +1,211 @@
+# Copyright 2019 XOE Corp. SAS (XOE Solutions)
+# License LGPL-3.0 or later (https://www.gnu.org/licenses/lgpl.html).
+
+from odoo import fields, models, _
+from odoo.exceptions import ValidationError
+
+
+VIEW_DOMAIN = "[('user_ids', '=', uid), ('company_id', '=', 'company_id')]"
+WITHOUT_COMPANY_ERROR = (
+    "An Operating Unit can only be set on records that also "
+    "have a company set (strict company namespacing)."
+    "Record {rec.display_name} has no company set."
+)
+NOT_SAME_COMPANY_ERROR = (
+    "The Operating Unit must be of the same company as the "
+    "record (company namespacing).Record {rec.display_name} has company "
+    "{rec.company_id.display_name} but it's operating unit {unit.display_name} "
+    "belongs to company {unit.company_id.display_name}."
+)
+
+TRANSACTION_REALM_BOUNDARIES_ERROR = (
+    "All bound metadata must share the operating unit of this transaction. "
+    "The record {rec.display_name} related on field {field.string} doesn't "
+    "share the operating unit of this transaction ({self.display_name})."
+)
+
+METADATA_NO_SHARED_REALM = (
+    "Linked and bound metadata must share at least one operating unit."
+    "The record {rec.display_name} related on field {field.string} doesn't "
+    "share an operating unit with this metadata ({self.display_name})."
+)
+
+class OpratingUnitRealmEnsurer(object):
+    """
+    Each record (metadata or transactional) knows two basic states, which are
+    enforced accordingly:
+     - "Bound" to one or more operating units
+     - "Unbound" without any operating unit (= globally available)
+
+    Bound records can be related to any unbound or to bound records which have
+    a common operating unit.
+
+    Unbound records can relate to any other record.
+    """
+
+    def _ensure_operating_unit_realm(self):
+        """ Ensures realm boundaries of operating units. All "participating"
+        records (transactional or metadata) must be either unbound or share at
+        least one common operating unit.
+
+        Yet an existing transaction, can passively be "expulsed" of a realm if
+        one of it's linked metadata's operating units are modified and would
+        top sharing one with the transaction.
+
+        "Expulsed" transaction block on write, unless their realm boundary is
+        healed or they are made unbound.
+
+        TODO: Ensure through SQL queries that transactions that would be
+        explused are exposed with a warning + confirmation (with the possibility
+        for the user to make them unbound)
```

Comment:
If that's not ensured there will be ugly leakages. And realms will be an incomplete design.

Reply:
Also check implication of potential metadata realm leakages.